File: sdks/dotnet/src/Rrflow.Rrd.Client/HttpTransport.cs
using System;
using System.Buffers;
using System.IO;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;

namespace Rrflow.Rrd;

internal sealed record HttpTransportResponse(int StatusCode, byte[] EncodedBody);

internal sealed class HttpTransport : IDisposable
{
    private readonly HttpClient _http;
    private readonly int _maxResponseBytes;

    internal HttpTransport(int maxResponseBytes, HttpMessageHandler? handler)
    {
        _maxResponseBytes = maxResponseBytes;
        _http = handler is null
            ? new HttpClient(new HttpClientHandler { AllowAutoRedirect = false })
            : new HttpClient(handler, disposeHandler: false);
    }

    internal async Task<HttpTransportResponse> SendAsync(
        HttpRequestMessage request,
        CancellationToken cancellationToken)
    {
        using HttpResponseMessage response = await _http.SendAsync(
            request,
            HttpCompletionOption.ResponseHeadersRead,
            cancellationToken).ConfigureAwait(false);
        byte[] encoded = await ReadBoundedAsync(response, cancellationToken).ConfigureAwait(false);
        return new HttpTransportResponse((int)response.StatusCode, encoded);
    }

    public void Dispose() => _http.Dispose();

    private async Task<byte[]> ReadBoundedAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        if (response.Content.Headers.ContentLength > _maxResponseBytes)
        {
            throw new RrdClientException("RRD response exceeded the configured byte limit");
        }
        await using Stream input = await response.Content.ReadAsStreamAsync(cancellationToken)
            .ConfigureAwait(false);
        using MemoryStream output = new();
        byte[] buffer = ArrayPool<byte>.Shared.Rent(8192);
        try
        {
            int total = 0;
            int count;
            while ((count = await input.ReadAsync(buffer, cancellationToken).ConfigureAwait(false)) > 0)
            {
                total += count;
                if (total > _maxResponseBytes)
                {
                    throw new RrdClientException("RRD response exceeded the configured byte limit");
                }
                output.Write(buffer, 0, count);
            }
            return output.ToArray();
        }
        finally
        {
            ArrayPool<byte>.Shared.Return(buffer);
        }
    }
}

File: sdks/java/src/main/java/io/rrflow/rrd/HttpTransport.java
package io.rrflow.rrd;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

/** Current HTTP carriage and response cap; H-07/J-02 own full resource policy. */
final class HttpTransport {
    record Response(int status, byte[] encoded) { }

    private final HttpClient http;
    private final int maxResponseBytes;

    HttpTransport(HttpClient http, int maxResponseBytes) {
        this.http = http;
        this.maxResponseBytes = maxResponseBytes;
    }

    Response send(HttpRequest request) throws IOException, InterruptedException {
        HttpResponse<InputStream> response =
                http.send(request, HttpResponse.BodyHandlers.ofInputStream());
        return new Response(response.statusCode(), readBounded(response));
    }

    private byte[] readBounded(HttpResponse<InputStream> response) throws IOException {
        long declared = response.headers().firstValueAsLong("Content-Length").orElse(-1);
        if (declared > maxResponseBytes) {
            throw new RrdClientException("RRD response exceeded the configured byte limit");
        }
        try (InputStream input = response.body();
                ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[8192];
            int total = 0;
            int count;
            while ((count = input.read(buffer)) != -1) {
                total += count;
                if (total > maxResponseBytes) {
                    throw new RrdClientException(
                            "RRD response exceeded the configured byte limit");
                }
                output.write(buffer, 0, count);
            }
            return output.toByteArray();
        }
    }
}
